day4/scratchcards.js

```diff
@@ -34,3 +32,72 @@ export function calculateWinnings(data) {
     })
     .reduce((a, b) => a + b);
 }
+
+/**
+ * Checks the current scratch card and provides the total number of matches
+ * @param {string[]} dataArr The raw string data array provided to us
+ * @param {int} idx The current index being analyzed
+ * @returns {int} The total number of matches found on the current line
+ */
+function sumMatches(dataArr, idx) {
+  let matches = 0;
+
+  const lineSplit = dataArr[idx].split("|");
+  const winningNumbers = lineSplit[0]
+    .split(":")[1]
+    .trim()
+    .split(" ")
+    .map((s) => parseInt(s))
+    .filter((n) => !isNaN(n));
+
+  const foundNumbers = lineSplit[1]
+    .trim()
+    .split(" ")
+    .map((s) => parseInt(s))
+    .filter((n) => !isNaN(n));
+
+  for (const num of foundNumbers) {
+    if (winningNumbers.find((n) => n === num)) {
+      matches++;
+    }
+  }
+
+  return matches;
+}
+
+/**
+ * Performs the recursion needed for the Day 4 Part 2 algorithm.
+ * @param {string[]} dataArr The raw data array.
+ * @param {int[]} totalArr The array keeping the tally of all of the cards.
+ * @param {int} idx The index we're currently on.
+ * @returns Nothing; returns when no more matches are found.
+ */
+function recurCalcTotal(dataArr, totalArr, idx) {
+  // termination condition
+  const matches = sumMatches(dataArr, idx);
```

Comment:
This is probably the best place to improve the performance of the algorithm. sumMatches is the biggest drain on resources, and we're running it once on every recursion. We really only want to run this once per line. Try starting the algorithm off by creating a new array containing the number of matches found on each line.

Reply:
Ok, this is fixed. Test suite was taking 30-60 seconds to run. Now runs in 60ms, which is plenty good enough.